src/combinatorics/hales_jewett.lean

```diff
@@ -0,0 +1,313 @@
+by { funext i, cases i; refl }
+
+@[simp] lemma prod_apply {α ι ι'} (l : line α ι) (l' : line α ι') (x : α) :
+  l.prod l' x = sum.elim (l x) (l' x) :=
+by { funext i, cases i; refl }
+
+@[simp] lemma diagonal_apply {α ι} [nonempty ι] (x : α) :
+  line.diagonal α ι x = λ i, x :=
+by simp_rw [line.apply, line.diagonal, option.get_or_else_none]
+
+/-- The Hales-Jewett theorem: for any finite types `α` and `κ`, there exists a finite type `ι` such
+that whenever the hypercube `ι → α` is `κ`-colored, there is a monochromatic combinatorial line. -/
+theorem exists_mono_in_high_dimension : ∀ (α : Type) [fintype α] (κ : Type) [fintype κ],
+  ∃ (ι : Type) [fintype ι], ∀ C : (ι → α) → κ, ∃ l : line α ι, l.is_mono C :=
+-- The proof proceeds by induction on `α`.
+fintype_induction
+-- We have to show that the theorem is invariant under `α ≃ α'` for the induction to work.
+(λ α α' e, forall_imp $ λ κ, forall_imp $ λ _, Exists.imp $ λ ι, Exists.imp $ λ _ h C,
+  let ⟨l, c, lc⟩ := h (λ v, C (e ∘ v)) in
+  ⟨l.map e, c, e.forall_congr_left.mp $ λ x, by rw [←lc x, line.map_apply]⟩)
+-- This deals with the degenerate case where `α` is empty.
+(λ κ hκ, ⟨κ, hκ, λ C, begin
+    haveI : nonempty κ := not_is_empty_iff.mp (λ ⟨h⟩, h (C $ λ i, (h i).elim)),
+    exact ⟨default _, classical.arbitrary κ, empty.rec _⟩
+  end⟩)
+-- The base case `α = unit` is easy since any line is monochromatic.
+(λ κ _, ⟨unit, infer_instance, λ C, ⟨diagonal _ _, C (λ _, ()), λ ⟨⟩, rfl⟩⟩)
+begin -- Now we have to show that the theorem holds for `option α` if it holds for `α`.
+  introsI α _ _ ihα κ _,
+-- The key idea is to show that for every `r`, in high dimension we can either find
+-- `r` color focused lines or a monochromatic line.
+  suffices key : ∀ r : ℕ, ∃ (ι : Type) [fintype ι], ∀ C : (ι → (option α)) → κ,
+    (∃ s : color_focused C, s.lines.card = r) ∨ (∃ l, is_mono C l),
+-- Given the key claim, we simply take `r = |κ| + 1`. We cannot have this many distinct colors so
+-- we must be in the second case, where there is a monochromatic line.
+  { obtain ⟨ι, _inst, hι⟩ := key (fintype.card κ + 1),
+    refine ⟨ι, _inst, λ C, (hι C).resolve_left _⟩,
+    rintro ⟨s, sr⟩,
+    apply nat.not_succ_le_self (fintype.card κ),
+    rw [←nat.add_one, ←sr, ←multiset.card_map, ←finset.card_mk],
+    exact finset.card_le_univ ⟨_, s.distinct_colors⟩ },
+-- We now prove the key claim, by induction on `r`.
+  intro r,
+  induction r with r ihr,
+-- The base case `r = 0` is trivial as the empty collection is color-focused.
+  { exact ⟨empty, infer_instance, λ C, or.inl ⟨default _, multiset.card_zero⟩⟩, },
+-- Supposing the key claim holds for `r`, we need to show it for `r+1`. First pick a high enough
+-- dimension `ι` for `r`.
+  obtain ⟨ι, _inst, hι⟩ := ihr,
+  resetI,
+-- Then since the theorem holds for `α` with and number of colors, pick a dimension `ι'` such that
+-- `ι' → α` always has a monochromatic line whenever it is colored with `(ι → α) → κ` colors.
```

Comment:
What do you mean by "`(ι → α) → κ` colors"? `|κ|^|α|^|ι|` colors? I don't quite get it.

Reply:
Yes this is confusing... I conflate finite types with natural numbers. I'll rewrite it. I mean that the type of colors is `(ι → α) → κ`. Normally in an informal presentation, you would probably have natural numbers `H, n, c` in place of these types, and then you'd have a coloring with `c^n^H` colors. But you would also use the equivalence `fin (c^n^H) = (fin H -> fin n) -> fin c`, so it's a bit easier to just work with the function type directly.

Reply:
Does it make more sense now? I try to consistently write "`X` is `Y`-colored" to mean we have a function `C : X -> Y`.

Reply:
Yeah, much better! You may want to explain this once in the module docstring but I think that's quite clear already.

Reply:
Great, I think we can leave it as it is then.